docs/partitions.md

```diff
@@ -0,0 +1,1208 @@
+- [Usage Examples](#usage-examples)
+
+---
+
+## Overview
+
+The partitions system in Gemini manages partition keys for database testing. It provides a thread-safe, memory-efficient way to:
+- Store and retrieve partition keys for testing
+- Track deleted partitions with time-based validation buckets
+- Generate new partitions dynamically
+- Replace existing partitions while maintaining consistency
+
+The system is crucial for Gemini's randomized testing approach, ensuring that mutations (INSERT, UPDATE, DELETE) and validations (SELECT) operate on a consistent set of partitions across both the system under test (SUT) and the test oracle.
+
+---
+
+## High-Level Architecture
+
+```
+┌─────────────────────────────────────────────────────────────┐
+│                     Partitions                              │
+│  ┌───────────────────────────────────────────────────────┐  │
+│  │                  Active Partitions                    │  │
+│  │  - Pre-allocated slice of partition key values       │  │
+│  │  - RW-mutex protected for concurrent access          │  │
+│  │  - Atomic counters (count, created)                  │  │
+│  └───────────────────────────────────────────────────────┘  │
+│                           │                                 │
+│                           │ Delete()                        │
+│                           ▼                                 │
+│  ┌───────────────────────────────────────────────────────┐  │
+│  │              Deleted Partitions Tracker               │  │
+│  │  ┌─────────────────────────────────────────────────┐  │  │
+│  │  │          Min-Heap (Priority Queue)             │  │  │
+│  │  │  - Sorted by readyAt time                      │  │  │
+│  │  │  - Pre-allocated backing array                 │  │  │
+│  │  │  - O(log n) insert, O(1) peek                  │  │  │
+│  │  └─────────────────────────────────────────────────┘  │  │
+│  │                      │                                 │  │
+│  │                      │ Background Processor            │  │
+│  │                      ▼                                 │  │
+│  │  ┌─────────────────────────────────────────────────┐  │  │
+│  │  │         Output Channel (buffered)              │  │  │
+│  │  │  - Sends ready partitions for validation      │  │  │
+│  │  │  - Buffered for throughput                     │  │  │
+│  │  └─────────────────────────────────────────────────┘  │  │
+│  └───────────────────────────────────────────────────────┘  │
+└─────────────────────────────────────────────────────────────┘
+```
+
+### Data Flow
+
+1. **Creation**: Partitions are pre-generated during initialization
+2. **Access**: Workers read partitions using distribution functions (uniform, zipfian, etc.)
+3. **Mutation**: Replace operations create new partition values
+4. **Deletion Tracking**: Old values are sent to the delete tracking system
+5. **Validation**: After time buckets expire, deleted partitions are emitted for validation
```

Comment:
is the same partition added back to tracking system next bucket? or we select bucket at deletion time?